kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Metadata.java

```diff
@@ -36,24 +36,30 @@
 import io.delta.kernel.internal.types.TableSchemaSerDe;
 
 public class Metadata {
-    public static Metadata fromRow(Row row, TableClient tableClient) {
-        if (row == null) {
+
+    public static Metadata fromColumnVector(
+            ColumnVector vector, int rowId, TableClient tableClient) {
+        if (vector.isNullAt(rowId)) {
             return null;
         }
 
-        final String schemaJson = requireNonNull(row, 4, "schemaString").getString(4);
+        final String schemaJson = requireNonNull(vector.getChild(4), rowId, "schemaString")
+            .getString(rowId);
         StructType schema = TableSchemaSerDe.fromJson(tableClient.getJsonHandler(), schemaJson);
 
         return new Metadata(
-            requireNonNull(row, 0, "id").getString(0),
-            Optional.ofNullable(row.isNullAt(1) ? null : row.getString(1)),
-            Optional.ofNullable(row.isNullAt(2) ? null : row.getString(2)),
-            Format.fromRow(requireNonNull(row, 0, "id").getStruct(3)),
+            requireNonNull(vector.getChild(0), rowId, "id").getString(rowId),
+            Optional.ofNullable(vector.getChild(1).isNullAt(rowId) ? null :
+                vector.getChild(1).getString(rowId)),
```

Comment:
what is the cost of `getChild`? Wondering with this approach are we creating too many objects?

Reply:
I think depends on the implementation; some create wrapper objects, we can update those implementations to only create the child vectors once.

It's no more object creation than the row API (and often less) since we'd create a wrapper row each `getStruct` call